File: infra-market-quarkus/src/main/kotlin/io/infra/qk/repository/PermissionRepository.kt
package io.infra.qk.repository

import io.infra.qk.entity.Permission
import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepository
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.persistence.EntityManager

/**
 * 权限 Repository
 * @author liuqinglin
 * Date: 2025/8/30
 */
@ApplicationScoped
class PermissionRepository @Inject constructor(
    private val entityManager: EntityManager
) : PanacheRepository<Permission> {
    
    /**
     * 根据ID查找权限
     */
    override fun findById(id: Long): Permission? {
        return find("id", id).firstResult()
    }
    
    /**
     * 根据权限编码查找权限
     */
    fun findByPermissionCode(permissionCode: String): Permission? {
        return find("code", permissionCode).firstResult()
    }
    
    /**
     * 根据父权限ID查找子权限
     */
    fun findByParentId(parentId: Long): List<Permission> {
        return find("parentId", parentId).list()
    }
    
    /**
     * 根据条件分页查询权限
     */
    fun findByCondition(permissionName: String?, permissionCode: String?, permissionType: String?, parentId: Long?, status: String?, page: Int, size: Int): Pair<List<Permission>, Long> {
        var query = "FROM Permission WHERE 1=1"
        val params = mutableMapOf<String, Any>()
        
        if (!permissionName.isNullOrBlank()) {
            query += " AND name LIKE :permissionName"
            params["permissionName"] = "%$permissionName%"
        }
        
        if (!permissionCode.isNullOrBlank()) {
            query += " AND code LIKE :permissionCode"
            params["permissionCode"] = "%$permissionCode%"
        }
        
        if (!permissionType.isNullOrBlank()) {
            query += " AND type = :permissionType"
            params["permissionType"] = permissionType
        }
        
        if (parentId != null) {
            query += " AND parentId = :parentId"
            params["parentId"] = parentId
        }
        
        if (!status.isNullOrBlank()) {
            query += " AND status = :status"
            params["status"] = status
        }
        
        query += " ORDER BY sort ASC, createTime DESC"
        
        val countQuery = query.replace("FROM Permission", "SELECT COUNT(*) FROM Permission")
        val total = entityManager.createQuery(countQuery, Long::class.java).apply {
            params.forEach { (key, value) -> setParameter(key, value) }
        }.singleResult
        
        val permissions = entityManager.createQuery(query, Permission::class.java).apply {
            params.forEach { (key, value) -> setParameter(key, value) }
            firstResult = (page - 1) * size
            maxResults = size
        }.resultList
        
        return Pair(permissions, total)
    }
    
    /**
     * 检查权限编码是否存在
     */
    fun existsByPermissionCode(permissionCode: String): Boolean {
        return count("code", permissionCode) > 0
    }
}

File: infra-market-server-go/internal/repository/permission_repository.go
package repository

import (
	"github.com/bucketheadv/infra-market/internal/dto"
	"github.com/bucketheadv/infra-market/internal/entity"
	"github.com/bucketheadv/infra-market/internal/util"
	"gorm.io/gorm"
)

type PermissionRepository struct {
	db *gorm.DB
}

func NewPermissionRepository(db *gorm.DB) *PermissionRepository {
	return &PermissionRepository{db: db}
}

// FindByCode 根据编码查询
func (r *PermissionRepository) FindByCode(code string) (*entity.Permission, error) {
	var permission entity.Permission
	err := r.db.Where("code = ? AND status != ?", code, "deleted").First(&permission).Error
	if err != nil {
		return nil, err
	}
	return &permission, nil
}

// FindByID 根据ID查询
func (r *PermissionRepository) FindByID(id uint64) (*entity.Permission, error) {
	var permission entity.Permission
	err := r.db.Where("id = ? AND status != ?", id, "deleted").First(&permission).Error
	if err != nil {
		return nil, err
	}
	return &permission, nil
}

// FindByIDs 批量查询
func (r *PermissionRepository) FindByIDs(ids []uint64) ([]entity.Permission, error) {
	if len(ids) == 0 {
		return []entity.Permission{}, nil
	}
	var permissions []entity.Permission
	err := r.db.Where("id IN ? AND status != ?", ids, "deleted").Find(&permissions).Error
	return permissions, err
}

// FindByParentID 根据父ID查询
func (r *PermissionRepository) FindByParentID(parentID uint64) ([]entity.Permission, error) {
	var permissions []entity.Permission
	err := r.db.Where("parent_id = ? AND status != ?", parentID, "deleted").Find(&permissions).Error
	return permissions, err
}

// FindAll 查询所有权限
func (r *PermissionRepository) FindAll() ([]entity.Permission, error) {
	var permissions []entity.Permission
	err := r.db.Where("status != ?", "deleted").Order("sort ASC, id ASC").Find(&permissions).Error
	return permissions, err
}

// Page 分页查询
func (r *PermissionRepository) Page(query dto.PermissionQueryDto) ([]entity.Permission, int64, error) {
	var permissions []entity.Permission

	db := r.db.Model(&entity.Permission{}).Where("status != ?", "deleted")
	db = ApplyCommonFilters(db, query.Name, query.Code, query.Status)
	
	if util.IsNotBlank(query.Type) {
		db = db.Where("type = ?", *query.Type)
	}

	return PaginateQuery(db, &query, "id ASC", &permissions)
}

// Create 创建权限
func (r *PermissionRepository) Create(permission *entity.Permission) error {
	return r.db.Create(permission).Error
}

// Update 更新权限
func (r *PermissionRepository) Update(permission *entity.Permission) error {
	return r.db.Save(permission).Error
}

// Delete 删除权限（软删除）
func (r *PermissionRepository) Delete(id uint64) error {
	return r.db.Model(&entity.Permission{}).Where("id = ?", id).Update("status", "deleted").Error
}

// Count 获取权限总数
func (r *PermissionRepository) Count() (int64, error) {
	var count int64
	err := r.db.Model(&entity.Permission{}).Where("status != ?", "deleted").Count(&count).Error
	return count, err
}

// CountBeforeDate 获取指定时间之前的权限总数
func (r *PermissionRepository) CountBeforeDate(timestamp int64) (int64, error) {
	var count int64
	err := r.db.Model(&entity.Permission{}).
		Where("status != ? AND create_time <= ?", "deleted", timestamp).
		Count(&count).Error
	return count, err
}

// FindByStatus 根据状态查询
func (r *PermissionRepository) FindByStatus(status string) ([]entity.Permission, error) {
	var permissions []entity.Permission
	err := r.db.Where("status = ?", status).Order("sort ASC, id ASC").Find(&permissions).Error
	return permissions, err
}
